8262291: Refactor reserve_memory_special_huge_tlbfs

kstefanj opened the pull request:
Please review this refactoring of the hugetlbfs reservation code.

**Summary**
In recent adventures in this area of the code I noticed a strange condition in `reserve_memory_special_huge_tlbfs` where we take the "mixed-mapping" route even if the size doesn't require any small pages to be used:
```
  if (is_aligned(bytes, os::large_page_size()) && alignment <= os::large_page_size()) {
    return reserve_memory_special_huge_tlbfs_only(bytes, req_addr, exec);
  } else {
    return reserve_memory_special_huge_tlbfs_mixed(bytes, alignment, req_addr, exec);
  }
```

The second condition here is needed because if the alignment is larger than the large page size, we needed to enforce this and can't just trust `mmap` to give us a properly aligned address. Doing this by using the mixed-function feels a bit weird and looking a bit more at this I found a way to refactor this function to avoid having the two helpers. 

Instead of only having the mixed path honor the passed down alignment, make sure that is always done. This will also have the side-effect that all large pages in a "mixed"-mapping will be at the start and then we will have a tail of small pages. This actually also ensures that we will use large pages for a mixed mapping, in the past there was a corner case where we could end up with just a head and tail of small pages and no large page in between (if the mapping was smaller than 2 large pages and there was no alignment constraint).

**Testing**
Mach5 tier1-3 and a lot of local testing with different large page configurations.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262291](https://bugs.openjdk.java.net/browse/JDK-8262291): Refactor reserve_memory_special_huge_tlbfs


### Reviewers
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer) ⚠️ Review applies to 38a131449c060eb6ebee32234f60356f64511cf1
 * @mgkwill (no known github.com user name / role) ⚠️ Review applies to 2bf1d5a59627313dc2a1dda70b573a4bc4ceb4f4
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3073/head:pull/3073` \
`$ git checkout pull/3073`

Update a local copy of the PR: \
`$ git checkout pull/3073` \
`$ git pull https://git.openjdk.java.net/jdk pull/3073/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3073`

View PR using the GUI difftool: \
`$ git pr show -t 3073`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3073.diff">https://git.openjdk.java.net/jdk/pull/3073.diff</a>

</details>

kstefanj commented:
/label remove hotspot-runtime
/label add hotspot-gc

kstefanj commented:
/label remove hotspot-gc
/label add hotspot

walulyai commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,98 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
-      ::munmap(lp_start, end - lp_start);
-      return NULL;
-    }
+  if (bytes == large_bytes) {
```
Shouldn't we do the check `if (large_mapping == NULL) {` before this?

kstefanj replied:
We could, and I thought about adding a comment here. I probably should have. The reason we don't do the explicit check it is that if `large_mapping == NULL` then NULL will be returned as expected and there is no additional work to be done if `bytes == large_bytes`.

walulyai replied:
When `bytes == large_bytes` don't we need to `unmap` the original reservation in case `reserve_and_commit_special`  fails i.e `large_mapping == NULL`? 

Below we have the comment 
// Large mapping failed, so we need to unmap the reminder
// of the orinal reservation.

Maybe I am missing something.

walulyai replied:
ok, now I get it. I had missed that when mmap fails, the original reservation is unmapped for this range so no need to do any additional clean up.

walulyai commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,98 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
+  // The requested size requires some small pages as well.
+  char* small_start = aligned_start + large_bytes;
+  size_t small_size = bytes - large_bytes;
+  if (large_mapping == NULL) {
```
See comment above

kstefanj replied:
I do the check here because I need the `small_start` and `small_size` values if `large_mapping == NULL` and there was additional memory reserved for small pages.

kstefanj commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,98 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
+  // be used and it must be aligned to both the large page size
+  // and the given alignment. The larger of the two will be used.
+  size_t required_alignment = MAX(os::large_page_size(), alignment);
+  char* const aligned_start = anon_mmap_aligned(req_addr, bytes, required_alignment);
```
My hope was that the new comment would be sufficient, but if you think it is needed I could add that the reserved range is not committed here.

kstefanj commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,98 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
-char* os::Linux::reserve_memory_special_huge_tlbfs_only(size_t bytes,
-                                                        char* req_addr,
-                                                        bool exec) {
+char* os::Linux::reserve_and_commit_special(size_t bytes,
```
That is a good point, I've struggled a bit with this function name. Since we actually do the reservation using the call to `anon_mmap_aligned()` maybe this one should just be called: `commit_memory_special()`

walulyai replied:
yes, seems better to me

kstefanj reviewed (commented):
Thanks for reviewing Ivan, I will wait to update the PR until we come up with a function name that is better than the current.

walulyai reviewed (approved):
lgtm!

mgkwill commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,100 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
+  if (large_mapping == NULL) {
+    // Failed to commit large pages, so we need to unmap the
+    // reminder of the orinal reservation.
+    ::munmap(small_start, small_size);
```
I'm assuming that if mmap fails for large pages, it un-maps the reservation area requested for large pages and thus here we only need to munmap for remaining reservation (small pages)?

kstefanj replied:
Yes, if `mmap()` fails we lose the reservation for that range and need to start over only using small pages.

mgkwill commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,100 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
-    return reserve_memory_special_huge_tlbfs_only(bytes, req_addr, exec);
-  } else {
-    return reserve_memory_special_huge_tlbfs_mixed(bytes, alignment, req_addr, exec);
+  // Commit the reminding bytes using small pages.
```
`  // Commit the reminding` should be `  // Commit the remaining`

mgkwill reviewed (approved):
Suggestions are mostly nits, if the assumption in my comment about un-mapping is correct (I so suspect based on your conversation with Ivan (@walulyai)). Looks good otherwise from my perspective.

kstefanj commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,100 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
-  int prot = exec ? PROT_READ|PROT_WRITE|PROT_EXEC : PROT_READ|PROT_WRITE;
-  int flags = MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED;
-  void* result;
+  // Start of by committing large pages.
```
Good catch, change this to `// First commit using large pages.` since I've already used "start off" above.

kstefanj commented:
> Suggestions are mostly nits, if the assumption in my comment about un-mapping is correct (I suspect so based on your conversation with Ivan (@walulyai)). Looks good otherwise from my perspective.

Thanks for reviewing Marcus, updated the comments per your suggestion and also changed one to better fit the flow.

mgkwill reviewed (approved):
+1

tstuefe reviewed (changes requested):
Hi Stefan,

this is a welcome cleanup!

Remarks inline.

Cheers, Thomas

kstefanj commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,100 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
+char* os::Linux::commit_memory_special(size_t bytes,
+                                       size_t page_size,
+                                       char* req_addr,
+                                       bool exec) {
```
That was my first plan as well, but it uses `os::Linux::hugetlbfs_page_size_flag()` which in turn depends on `os::Linux::default_large_page_size()`. So I went with this. 

I agree that we long term should look at using/merging `commit_memoy()` but didn't want to change to much functionality at once here. What we get now is very much a like what we got before but all large pages at the beginning of the mapping. 

I also agree that now when this is really just a commit function, returning the address is a bit strange. I don't see a good way to handle potential failures in here so I opted for returning a `bool`.

In some sense it is left to the caller (`ReservedSpace`) right now. If using large pages fail, we will try using small pages instead. But not sure if you would want this logic to move to another layer.

kstefanj commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -3918,139 +3918,100 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
+char* os::Linux::reserve_memory_special_huge_tlbfs(size_t bytes,
+                                                   size_t alignment,
+                                                   char* req_addr,
+                                                   bool exec) {
```
Yes, right now we will only use a single large page size at a time. But using more than one (after Marcus change) need to take the alignment into account. As you know I have plans to do more changes in this area so that we can pass down what page size we want for a given reservation. But if we want to apply that change before everything else is in place I will suggest to use the alignment as an upper limit for the page size. 

A mixed mapping would still only use one large page size and then fill up with small pages. This is not optimal, but trying to mix and match even more will lead to a lot more error handling and if doing that it should be a separate change.

For cases where we only expect to use large pages (the old only case) the alignment is always set to `large_page_size` or larger.

kstefanj commented:
Thanks for reviewing @tstuefe, if you are ok with these changes I will push this tonight as I will be on parental leave for the coming two weeks.

kstefanj commented:
@tstuefe, I'm back from my leave now. Do you have any additional comments?

tstuefe commented:
Looks good to me.

Thanks, Thomas

kstefanj commented:
Thanks @tstuefe, you have to actively approve the PR to make it ready for integration since you are the only Reviewer.

tstuefe reviewed (approved):
Oh right, sorry.

Looks good Stefan!

..Thomas

kstefanj commented:
Thanks for the reviews @tstuefe, @mgkwill and @walulyai

/integrate
